### PR TITLE
CAD-4648 Document that protocol versions for eras do not have to be successive.

### PR DESCRIPTION
**Pull request by jorisdral**
A recent bug in `db-analyser` was caused by misconfiguring the era transition from Alonzo to Babbage: It was set to be triggered at protocol version 6, while it should be triggered at protocol version 7. This implies that we can skip protocol versions, i.e., protocol version 6. This PR aims to document that:
* From the view of consensus, protocol versions do not have to be successive.
* The Babbage hard-fork should be triggered at protocol version 7, and not 6.

# Checklist

- Branch
    - [x] Commit sequence broadly makes sense
    - [x] Commits have useful messages
    - [x] New tests are added if needed and existing tests are updated
    - [x] If this branch changes Consensus and has any consequences for downstream repositories or end users, said changes must be documented in [`interface-CHANGELOG.md`](../ouroboros-consensus/docs/interface-CHANGELOG.md)
    - [x] If this branch changes Network and has any consequences for downstream repositories or end users, said changes must be documented in [`interface-CHANGELOG.md`](../docs/interface-CHANGELOG.md)
    - [x] If serialization changes, user-facing consequences (e.g. replay from genesis) are confirmed to be intentional.
- Pull Request
    - [x] Self-reviewed the diff
    - [x] Useful pull request description at least containing the following information:
      - What does this PR change?
      - Why these changes were needed?
      - How does this affect downstream repositories and/or end-users?
      - Which ticket does this PR close (if any)? If it does, is it [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)?
    - [x] Reviewer requested

### CODE REVIEW

File: ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/ShelleyHFC.hs [hunk 1]
```diff
@@ -177,6 +177,13 @@ instance
     LedgerSupportsProtocol (ShelleyBlock proto era)
   ) => SingleEraBlock (ShelleyBlock proto era) where
   singleEraTransition pcfg _eraParams _eraStart ledgerState =
+      -- TODO(jdral): When fixing a bug in @db-analyser@ related to hard-forks,
+      -- we replaced the matched expression below with @traceShowId $
+      -- shelleyTriggerHardFork pcf@ to debug the hard-fork functionality.
+      -- The trace of @db-analyser@ shows that the expression is printed thrice
+      -- for every @reapplyBlock@ of a Shelley block. Does this mean that we
+      -- evaluate this function thrice for every @reapplyBlock@? It would be
+      -- useful to investigate whether this is problematic behaviour.
```
Comment (jorisdral) [inline ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/ShelleyHFC.hs:186]: This will print something along the lines of the following:
```
TriggerHardForkAtVersion 3
TriggerHardForkAtVersion 3
TriggerHardForkAtVersion 3
```

**Review by Jasagredo — APPROVED** [review]
This looks good to me. I would probably also make a note [here](https://github.com/input-output-hk/ouroboros-network/blob/master/ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/Node.hs#L546) about it being `7` instead of `6`

File: ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Simple.hs [hunk 2]
```diff
@@ -13,6 +13,10 @@ import           Cardano.Slotting.Slot (EpochNo)
 data TriggerHardFork =
     -- | Trigger the transition when the on-chain protocol major version (from
     -- the ledger state) reaches this number.
+    --
+    -- Note: Consensus does not enforce protocol versions to be successive.
+    -- That is, the ledger state is not restricted to transition from version
+    -- @x@ to @x+1@.
```
[the remark this answers is not in the document]
Comment (dnadales) [reply ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Simple.hs:19]: I pinged @nfrisby.

**Review by dnadales — COMMENTED** [review]
Thanks! LGTM. Only a couple of comments to address.

File: ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Simple.hs (hunk 2 above)
[the remark this answers is not in the document]
Comment (nfrisby) [reply ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Simple.hs:19]: Seems like a reasonable place to me 👍

[the remark this answers is not in the document]
Comment (nfrisby) [reply ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Simple.hs:19]: Err. I'd rephrase it a bit though.

The ledger rules determine how the state is allowed to change (and the Shelley ledger eg might require it can only increase by one, contradicting the current phrasing of your comment). But Consensus makes no assumptions about it -- I suppose we do assume that it never goes down...

I think the more precise comment is that the HFC logic does not require that the trigger version for one era is the successor of the trigger version for the previous era.

File: ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/ShelleyHFC.hs (hunk 1 above)
Comment (jorisdral) [reply ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/ShelleyHFC.hs]: I've created ticket CAD-4666, which also links to the updated TODO in the code.
